File: frontend/src/types/inspect.ts
export type StepState = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED';

export interface StepRecord {
  step_record_id: string;
  logical_run_id: string;
  run_id: string;
  run_kind: string;
  attempt_number: number;
  step_name: string;
  step_index: number;
  state: StepState;
  project_id: string;
  model_id?: string;
  backend_name?: string;
  started_at?: string;
  finished_at?: string;
  duration_seconds?: number;
  error_code?: string;
}

export interface InspectContext {
  jobId: string;
  runKind: 'pipeline_job' | 'role_model_check';
  attemptNumber?: number;
}

export type ArtifactState = 'CANONICAL' | 'SUPERSEDED' | 'REJECTED' | 'DRAFT';

export type ArtifactKind = 
  | 'PROJECT_BRIEF'
  | 'CHAPTER_PLAN'
  | 'SEQUENCE'
  | 'SCENE_STORYBOARD'
  | 'STORY_BIBLE'
  | 'MANUSCRIPT_DOCUMENT';

export interface ArtifactLineageView {
  artifact_id: string;
  artifact_kind: ArtifactKind;
  state: ArtifactState;
  run_id: string;
  step_name: string;
  created_at: string;
  provenance?: {
    provider?: string;
    model?: string;
    backend_name?: string;
  };
}

export interface AttemptHistoryItem {
  attempt_number: number;
  status: string;
  executor_name: string | null;
  executor_instance_id: string | null;
  queue_delay_ms: number | null;
  lease_owner: string | null;
  lease_expires_at: string | null;
  claimed_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  last_heartbeat_at: string | null;
  finish_reason: string | null;
  failure_stage: string | null;
  retryable: boolean | null;
  retry_reason: string | null;
  error_code: string | null;
  error_category: string | null;
}

export interface JobAttemptHistoryResponse {
  job_id: string;
  items: AttemptHistoryItem[];
  meta: Record<string, string>;
}

export interface RoleModelCheckAttemptHistoryResponse {
  run_id: string;
  items: AttemptHistoryItem[];
  meta: Record<string, string>;
}

File: app/schemas/inspect.py
from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import Field

from .base import StrictModel


class StepRecordView(StrictModel):
    step_record_id: int
    logical_run_id: str
    run_id: str
    run_kind: str
    attempt_number: int
    step_name: str
    step_index: int
    state: str
    project_id: str | None = None
    model_id: str | None = None
    critic_profile: str | None = None
    backend_name: str | None = None
    backend_version: str | None = None
    input_hash: str | None = None
    output_hash: str | None = None
    prompt_hash: str | None = None
    input_artifact_refs: list[str] = Field(default_factory=list)
    output_artifact_refs: list[str] = Field(default_factory=list)
    started_at: str | None = None
    finished_at: str | None = None
    duration_seconds: float | None = None
    finish_reason: str | None = None
    error_code: str | None = None
    error_category: str | None = None
    executor_id: str | None = None
    lease_owner: str | None = None


class ArtifactLineageView(StrictModel):
    artifact_lineage_id: int
    logical_run_id: str
    run_id: str
    run_kind: str
    attempt_number: int
    step_name: str
    project_id: str | None = None
    artifact_role: str
    artifact_kind: str
    path: str
    content_hash: str | None = None
    status: str
    validation_state: str
    produced_at: str
    registered_at: str | None = None
    supersedes_artifact_lineage_id: int | None = None
    source_artifact_refs: list[str] = Field(default_factory=list)
    source_content_hashes: list[str] = Field(default_factory=list)
    output_of_step_record_id: int


class JobStepsResponse(StrictModel):
    job_id: UUID
    items: list[StepRecordView] = Field(default_factory=list)
    meta: dict[str, Any] = Field(default_factory=lambda: {"ordered_by": "step_index_asc"})


class JobLineageResponse(StrictModel):
    job_id: UUID
    items: list[ArtifactLineageView] = Field(default_factory=list)
    meta: dict[str, Any] = Field(default_factory=lambda: {"ordered_by": "artifact_lineage_id_asc"})


class RoleModelCheckStepsResponse(StrictModel):
    run_id: UUID
    items: list[StepRecordView] = Field(default_factory=list)
    meta: dict[str, Any] = Field(default_factory=lambda: {"ordered_by": "step_index_asc"})


class RoleModelCheckLineageResponse(StrictModel):
    run_id: UUID
    items: list[ArtifactLineageView] = Field(default_factory=list)
    meta: dict[str, Any] = Field(default_factory=lambda: {"ordered_by": "artifact_lineage_id_asc"})


class AttemptHistoryItem(StrictModel):
    attempt_number: int
    status: str
    executor_name: str | None = None
    executor_instance_id: str | None = None
    queue_delay_ms: int | None = None
    lease_owner: str | None = None
    lease_expires_at: str | None = None
    claimed_at: str | None = None
    started_at: str | None = None
    finished_at: str | None = None
    last_heartbeat_at: str | None = None
    finish_reason: str | None = None
    failure_stage: str | None = None
    retryable: bool | None = None
    retry_reason: str | None = None
    error_code: str | None = None
    error_category: str | None = None


class JobAttemptHistoryResponse(StrictModel):
    job_id: UUID
    items: list[AttemptHistoryItem] = Field(default_factory=list)
    meta: dict[str, Any] = Field(default_factory=lambda: {"ordered_by": "attempt_number_asc"})


class RoleModelCheckAttemptHistoryResponse(StrictModel):
    run_id: UUID
    items: list[AttemptHistoryItem] = Field(default_factory=list)
    meta: dict[str, Any] = Field(default_factory=lambda: {"ordered_by": "attempt_number_asc"})
